File: compiler/src/Gallery.hs
-- ldgallery - A static generator which turns a collection of tagged
--             pictures into a searchable web gallery.
--
-- Copyright (C) 2019  Pacien TRAN-GIRARD
--
-- This program is free software: you can redistribute it and/or modify
-- it under the terms of the GNU Affero General Public License as
-- published by the Free Software Foundation, either version 3 of the
-- License, or (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU Affero General Public License for more details.
--
-- You should have received a copy of the GNU Affero General Public License
-- along with this program.  If not, see <https://www.gnu.org/licenses/>.

{-# LANGUAGE
    DuplicateRecordFields
  , DeriveGeneric
  , DeriveAnyClass
#-}

module Gallery
  ( GalleryItem(..), buildGallery
  ) where


import GHC.Generics (Generic)
import Data.Char (toLower)
import Data.Function ((&))
import Data.Maybe (fromMaybe)

import Data.Aeson (ToJSON, genericToJSON, genericToEncoding)
import qualified Data.Aeson as JSON

import qualified Data.Set as Set

import Files
import Input
import Resource


encodingOptions :: JSON.Options
encodingOptions = JSON.defaultOptions
  { JSON.fieldLabelModifier = map toLower
  , JSON.constructorTagModifier = map toLower
  , JSON.sumEncoding = JSON.defaultTaggedObject
      { JSON.tagFieldName = "type"
      , JSON.contentsFieldName = "contents"
      }
  }


type ResourcePath = String
type Tag = String
type FileSizeKB = Int


data Resolution = Resolution
  { width :: Int
  , height :: Int
  } deriving (Generic, Show)

instance ToJSON Resolution where
  toJSON = genericToJSON encodingOptions
  toEncoding = genericToEncoding encodingOptions


data GalleryItemProps =
    Directory { items :: [GalleryItem] }
--  | Image { resolution :: Resolution, filesize :: FileSizeKB }
--  | Video { filesize :: FileSizeKB }
  | Unknown
  deriving (Generic, Show)

instance ToJSON GalleryItemProps where
  toJSON = genericToJSON encodingOptions
  toEncoding = genericToEncoding encodingOptions


-- TODO: fuse GalleryItem and GalleryItemProps
data GalleryItem = GalleryItem
  { title :: String
  , date :: String -- TODO: checked ISO8601 date
  , description :: String
  , tags :: [Tag]
  , path :: ResourcePath
  , thumbnail :: Maybe ResourcePath
  , properties :: GalleryItemProps
  } deriving (Generic, Show)

instance ToJSON GalleryItem where
  toJSON = genericToJSON encodingOptions
  toEncoding = genericToEncoding encodingOptions


buildGalleryTree :: ResourceTree -> GalleryItem
buildGalleryTree (ItemResource sidecar path@(filename:_) thumbnail) =
  GalleryItem
    { title = optMeta title filename
    , date = optMeta date "" -- TODO: check and normalise dates
    , description = optMeta description ""
    , tags = optMeta tags []
    , path = webPath path
    , thumbnail = fmap webPath thumbnail
    , properties = Unknown } -- TODO
  where
    optMeta :: (Sidecar -> Maybe a) -> a -> a
    optMeta get fallback = fromMaybe fallback $ get sidecar

buildGalleryTree (DirResource dirItems path@(dirname:_) thumbnail) =
  map buildGalleryTree dirItems
  & \items -> GalleryItem
    { title = dirname
      -- TODO: consider using the most recent item's date? what if empty?
    , date = ""
      -- TODO: consider allowing metadata sidecars for directories too
    , description = ""
    , tags = aggregateChildTags items
    , path = webPath path
    , thumbnail = fmap webPath thumbnail
    , properties = Directory items }
  where
    aggregateChildTags :: [GalleryItem] -> [Tag]
    aggregateChildTags = unique . concatMap (\item -> tags (item::GalleryItem))

    unique :: Ord a => [a] -> [a]
    unique = Set.toList . Set.fromList

buildGallery :: String -> ResourceTree -> GalleryItem
buildGallery galleryName resourceTree =
  (buildGalleryTree resourceTree) { title = galleryName }

File: viewer/src/@types/gallery.ts
/* ldgallery - A static generator which turns a collection of tagged
--             pictures into a searchable web gallery.
--
-- Copyright (C) 2019-2022  Guillaume FOUET
--
-- This program is free software: you can redistribute it and/or modify
-- it under the terms of the GNU Affero General Public License as
-- published by the Free Software Foundation, either version 3 of the
-- License, or (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU Affero General Public License for more details.
--
-- You should have received a copy of the GNU Affero General Public License
-- along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

import { ItemType } from './itemType';
import { SplashScreenConfig } from './splashscreen';

export type ItemSortStr = 'title_asc' | 'date_asc' | 'date_desc';

export interface Config {
  galleryRoot: string;
  galleryIndex?: string;
  initialItemSort?: ItemSortStr;
  initialTagDisplayLimit?: number;
  splashScreen?: SplashScreenConfig;
}

// ---

export interface Resolution {
  width: number;
  height: number;
}
export interface Thumbnail {
  resource: string;
  resolution: Resolution;
}
export type RawTag = string;

// ---

export interface Downloadable {
  resource: string;
  type: ItemType; // unknown
}
export interface OtherProperties extends Downloadable {
  type: ItemType.OTHER;
}
export interface PictureProperties extends Downloadable {
  type: ItemType.PICTURE;
  resolution: Resolution;
}
export interface PlainTextProperties extends Downloadable {
  type: ItemType.PLAINTEXT;
}
export interface MarkdownProperties extends Downloadable {
  type: ItemType.MARKDOWN;
}
export interface PDFProperties extends Downloadable {
  type: ItemType.PDF;
}
export interface VideoProperties extends Downloadable {
  type: ItemType.VIDEO;
}
export interface AudioProperties extends Downloadable {
  type: ItemType.AUDIO;
}
export interface DirectoryProperties {
  type: ItemType.DIRECTORY;
  // eslint-disable-next-line no-use-before-define
  items: Item[];
}

// ---

export interface Item {
  title: string;
  datetime: string;
  description: string;
  tags: RawTag[];
  path: string;
  thumbnail?: Thumbnail;
  properties:
    | Downloadable
    | DirectoryProperties;
}
export interface DownloadableItem extends Item { // Special unknown item type
  properties: Downloadable;
}
export interface OtherItem extends Item {
  properties: OtherProperties;
}
export interface PictureItem extends Item {
  properties: PictureProperties;
}
export interface PlainTextItem extends Item {
  properties: PlainTextProperties;
}
export interface MarkdownItem extends Item {
  properties: MarkdownProperties;
}
export interface PDFItem extends Item {
  properties: PDFProperties;
}
export interface VideoItem extends Item {
  properties: VideoProperties;
}
export interface AudioItem extends Item {
  properties: AudioProperties;
}
export interface DirectoryItem extends Item {
  properties: DirectoryProperties;
}

// ---

export interface IndexProperties {
  galleryTitle: string;
  tagCategories: RawTag[];
}
export interface Index {
  properties: IndexProperties;
  tree: DirectoryItem;
}
